src/screen.h

```diff
@@ -192,4 +194,60 @@ bool screen_force_clear_to_end();
 /// Returns the length of an escape code. Exposed for testing purposes only.
 size_t escape_code_length(const wchar_t *code);
 
+// Maintain a mapping of escape sequences to their length for fast lookup.
+class cached_esc_sequences_t {
```

Comment:
Clever approach to cache the lengths and the escape sequences separately, and just iterate substrings of each length. Much simpler than what I thought of!